### ISSUE TITLE
Link field work and improvements

### ISSUE THREAD

**Issue by JoryHogeveen** [pods-framework/pods#2655]
Is it possible to add a Link field to Pods?
Would be a bit similar to a Website field I think.
- URL (possible with the Link feature from wordpress itself so you can select the page to link to.
- Title
- Target

Could be separate metadata in the database or a serialized array.

**Last items for Pods 2.7** #4068 
- [x] What to do if the user changes the field type from Website to Link and vice versa?
See: https://github.com/JoryHogeveen/pods-link-field/blob/master/classes/fields/link.php#L252-L255
- [x] Add an option to set the default target value to _blank
See: JoryHogeveen/pods-link-field@5a07ddd
- [x] @sc0ttkclark Do you want to enable other storage types as well? It currently only stores as a serialized array

**Comment by Shelob9** [comment]
We are not adding any new features to Pods 2.X. Once we finish the cmb2 integration for Pods 3.0 we can address new field types. This would be one be very nice to have, possibly as a specific type of loop field.

**Comment by JoryHogeveen** [comment]
Is there a guide/howto on creating extra field types? I would like to try and start with this field on the 3.0 release.

**Comment by sc0ttkclark** [comment]
If you can start with building the frontend for it and how it would work / look (using the same aesthetics as WP), I can meet you in the middle and help you adjust the current website field to support this field format.

**Comment by JoryHogeveen** [comment]
Hi Scott, I assume you mean wireframes and a HTML version of the a link.php template (ui/fields/link.php)?

**Comment by sc0ttkclark** [comment]
Yes, I'd like to see some frontend markup and styling for how you envision it looking.

**Comment by JoryHogeveen** [comment]
![pods link field front-end](https://cloud.githubusercontent.com/assets/826148/6305103/18f9aba6-b929-11e4-9626-64f11f87bde7.png)

This is how I would see this working. Especially the popup feature would be neat.

HTML version for the file link.php in ui/fields:

```html
<div id="link-options">
    <p class="howto"><?php _e('Enter the destination URL') ?></p>
    <div class="alignleft">
        <label><span><?php _e('URL') ?></span><input id="url-field" type="text" name="href"></label>
    </div>
    <div class="alignleft">
        <label><span><?php _e('Title') ?></span><input id="link-title-field" type="text" name="linktitle"></label>
    </div>
    <div class="link-target">
        <label><div>&nbsp;</div><input type="checkbox" id="link-target-checkbox" style="display: inline-block;"> <?php _e('Open link in a new window/tab') ?></label>
    </div>
    <p class="howto" style="clear: both;"><a href="#" id="wp-link-search-toggle"><?php _e('Or link to existing content') ?></a> (Popup WYSIWYG Link feature &amp; autofill above fields)</p>
</div>
```

**Comment by JoryHogeveen** [comment]
I've searched some extra info on the WYSIWYG link feature. Is done with a javascript class (wp-includes / js / wplink.js) and i'm afraid its build for the editor only. We could use this as a blueprint for this field...

**Comment by JoryHogeveen** [comment]
An class on this feature: wp-includes / class-wp-editor.php

lines:
- 1231 - 1232
- 1434 - 1499 ( wp_link_dialog() function )

**Comment by sc0ttkclark** [comment]
Great research here, would love to find ways to integrate into the link to existing and figure out how to utilize all of this on non post editor areas like user edit / taxonomy term add/edit, or even on the frontend of a site.

**Comment by sc0ttkclark** [comment]
UI of the field looks great to me by the way

**Comment by quasel** [comment]
+1 looks good

**Comment by Dooza** [comment]
Just wondering how this one was progressing as I have a need for it.

**Comment by JoryHogeveen** [comment]
It doesn't have a contributor yet as I can see so far.

Scott, I would like to help in the development on this though I think I will need some help from a Pods developer.
Can I find some quidelines for developing in Pods? And is the latest version on GitHub stable enough for development and testing?

**Comment by JoryHogeveen** [comment]
Also, here is a plugin for ACF witch does a simpler version of what we want to do:
https://wordpress.org/plugins/advanced-custom-fields-link-picker-field/

**Comment by Dooza** [comment]
That is exactly how it should look, no point reinventing the wheel.

**Comment by Dooza** [comment]
Not sure if this is any help: https://wordpress.org/plugins/link-chooser/

**Comment by JoryHogeveen** [comment]
My UI above wasn't adds the option to manually enter the URL + it adds the option to specify a tekst for the link (instead of just showing the URL in the front end) and the option to open the link in a new tab.
I think these are important features to have when using a Link field type.

The Link Chooser plugin will definately help I think, thanks for sharing!

**Comment by JoryHogeveen** [comment]
Well, then I think there is only one think left.

Scott, how would you like to store this in the post_meta table? I think a serialized array would be good like this:

array(
'url' => THE_URL,
'title' => THE_TITLE,
'target' => BOOLEAN (0 is same tab, 1 is new tab)
)

**Comment by quasel** [comment]
if you are going to work on this take a look at https://github.com/WebDevStudios/CMB2/wiki/Adding-your-own-field-types - Pods 3.0 will integrate CMB2 has support for it pods will too

**Comment by JoryHogeveen** [comment]
Thanks for the info. Though I understood that this would be a core field for Pods. Will it then still be implemented this way?

**Comment by sc0ttkclark** [comment]
I think it'd be great to have a link field type built into Pods, so I think we'd roll it in.

I'd like to see a link field that makes use of the WP Editor Link modal popup and supports title / link / target. It'd store as {field_name}, {field_name}_title, {field_name}_target probably, for meta-based storage. I'm not opposed to storing it serialized though.

**Comment by JoryHogeveen** [comment]
Great! I think serialized would be more than sufficiënt, plus this way it only requires one row in the db table.
What are your thoughts? And what would the benefits be of storing this separate?

**Comment by JoryHogeveen** [comment]
Hi all!
I've finally found the time to continue this.
Current status (development is done ins Pods 2.6):

- [x] Storage in the meta table (serialized array)
- [x] Use the wpLink modal for selecting existing pages and let it update the fields.
- [x] Link formats: Same as "website" field but added the "none" options so you can also add "mailto:" or "tel:" links.
- [x] optionally allow HTML in the title field (same functionality as the "text" field).
- [] Not sure: max length for table storage
- [] Enable use of shortcodes in the text and url input fields (could be handy sometimes). The option in the admin UI works. only the front-end display is not working yet. See next item...
- [] Display function: the PodsForm class only sends one array value at a time, need to find a workaround to get all link-field values to enable rendering of a link in HTML. I'm still figuring out how the Pods system works on rendering fields to the front-end. (for example: also found a "raw" option..)

So.. the field works, but at this point it is only possible to get the field as an array through get_post_meta. Not as a HTML link (pods display/view/field functions are not working yet).

Help on the Pods functions would be nice!

**Comment by JoryHogeveen** [comment]
@sc0ttkclark 
using pods_field_display works great now. One question:

There are two more functions: pods_field and pods_field_raw.
pods_field_raw obviously returns the database data but what shall I return with pods_field?

**Comment by JoryHogeveen** [comment]
@jimtrue 
Maybe handy for Pods docs:

Data storage (and return) when chosing raw is this:

```php
Array
(
    [url] => val
    [text] => val
    [target] => val
)
```

**Comment by sc0ttkclark** [comment]
Duplicate of the original #708

**Comment by JoryHogeveen** [comment]
Reopening this for two reasons:

- What to do if the user changes the field type from `Website` to `Link`?
See: https://github.com/JoryHogeveen/pods-link-field/blob/master/classes/fields/link.php#L252-L255

- Add an option to set the default target value to `_blank`
See: https://github.com/JoryHogeveen/pods-link-field/commit/5a07ddd70c54071dbbac2554c432b31019ffa98b

**Comment by sc0ttkclark** [comment]
@JoryHogeveen sorry for not getting back to you. Do you want to tackle these things in 2.7 or can we push them to later?

**Comment by JoryHogeveen** [comment]
@sc0ttkclark I'd like to tackle these for 2.7 since I believe it will raise a lot of support questions otherwise.
Also.. I already have it fixed for the most part so why not...

The last topic is your area so I'll leave that up to you.

**Comment by jimtrue** [comment]
Never actually tested this one, sorry. Output isn't correct in a Magic Tag (ie we should just be able to go `{@link_field}` and it do all the necessary wrapping, etc.

this is what it outputs: `and _blank`

I think because it's actually built as an array, it output it as a display() list, so url, text, and target. It should actually _construct_ the link.

**Comment by jimtrue** [comment]
ie, it should have a special handler for display() like your output as gallery option.

**Comment by quasel** [comment]
While this is a neat idea it might lead to confused users - just adding a field should I use link or Website? whats the difference? In my opinion this two fields need to be combined and act like the image stuff to be consistent! 

- needs magic tag support maybe including stuff like for images _src / _schemeless 
- needs display() support 

@sc0ttkclark @JoryHogeveen - do you want another issue? for the time being I just reopened this one!

**Comment by sc0ttkclark** [comment]
I think you're right, we need to differentiate these two field types much better in describing/naming them. As far as combining them, there's some technical issues to overcome, one is a text input and the other has a few different parts to it.

Magic tag / display support could probably be improved too, I'm open to potentially putting this new field type inclusion on hold until we get the naming thing sorted out. We could comment out the `link` field type inclusion pretty easily in PodsForm.

**Comment by quasel** [comment]
At least until the output for dispaly and magic tags works correctly we should not put it out in the wild - i like the new approach and i would see it as a superior solution - maybe we can convert the old field to it and add a option in the configuration? something similar to how images work or we add another group of "URL" Kind of fields ... 

@reachadam from WPCrafter already asked me about the weird output ;)

**Comment by JoryHogeveen** [comment]
Let me know if you have any ideas for combining these fields. With these changes I think renaming the Website field is mandatory since I don't think it described exactly what is done here.
We could just rename the label to "Website / Link" and merge all the Link functionality into the Website field class with an option to enable/disable it?

**Comment by jimtrue** [comment]
My thought would be if you checked the 'link' button, that's when the additional options would show up because then they would be needed. Typically the website field would be designed to contain the URL portion of the 'link'. 

I've always felt the Website field is actually a URL field. I use it for that with YouTube links, Slideshare, etc. The old 'link' post type that was part of WordPress was designed more like you've set this one up to be. If we were to combine them, the other options should only show if you want the 'link' to be handled automatically with `echo $pods->display('url_field');` or `{@url_field}`, which ideally, display and magic tag display should be identically handled. The only time it would return the array setup would be if you did a `$pods->field('url_field');`

**Comment by quasel** [comment]
Idea / Summary:
- link_field display and magic tag of the field per default returns the url as configured - Like the "Gallery" Feature you can opt in to output a full "link"
- link_field._src just the plain url
- link_field._src_schemeless
- link_field._src_relative
and for the full output? _link or _url or ^^^^ like it's done for images and files
- link_field._link always puts out a full link if no "text" is configured the url is used as text too and _blank
...

maybe not the best approach but seems inline with the current options for files / images / and it should work for "old" and "new" thoughts?

**Comment by pglewis** [comment]
Is this still planned to make 2.7?

**Comment by sc0ttkclark** [comment]
Do we have a list of todos left on this or are the remaining things left just nice-to-haves? Should we punt to 2.8?

**Comment by JoryHogeveen** [comment]
I think we'll need to clearify what to do exactly first before we can create a todo list.
If 2.7 will release on aug 15th as planned according to the milestone I think it's better to punt to 2.8 or a 2.7 maintenance release

**Comment by JoryHogeveen** [comment]
New request/vote for this field: https://wordpress.org/support/topic/custom-fields-to-store-link-and-its-description/
